extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -55,12 +69,49 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		this._onReadAllTOCFiles.fire();
 	}
 
+	async openBook(bookPath: string, context: vscode.ExtensionContext, openAsUntitled: boolean): Promise<void> {
+		try {
+			// Check if the book is already open in viewlet.
+			if (this._tableOfContentPaths.indexOf(path.join(bookPath, '_data', 'toc.yml').replace(/\\/g, '/')) > -1 && this._allNotebooks.size > 0) {
+				vscode.commands.executeCommand('workbench.books.action.focusBooksExplorer');
+			}
+			else {
+				await this.getTableOfContentFiles([bookPath]);
+				let bookViewer = vscode.window.createTreeView('bookTreeView', { showCollapseAll: true, treeDataProvider: this });
+				vscode.commands.executeCommand('workbench.books.action.focusBooksExplorer').then(res => {
+					this._openAsUntitled = openAsUntitled;
+					let books = this.getBooks();
+					if (books && books.length > 0) {
+						bookViewer.reveal(books[0], { expand: vscode.TreeItemCollapsibleState.Expanded, focus: true, select: true });
+						const readmeMarkdown: string = path.join(bookPath, 'content', books[0].tableOfContents[0].url.concat('.md'));
+						const readmeNotebook: string = path.join(bookPath, 'content', books[0].tableOfContents[0].url.concat('.ipynb'));
+						if (fs.existsSync(readmeMarkdown)) {
```

Comment:
Don't use the sync methods - instead wrap the async versions using promisify https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original